src/App.tsx

```diff
@@ -1,24 +1,215 @@
+        }));
+      })
+      .catch(() => {
+        setError(ErrorMessages.LoadFail);
+      });
+  }, []);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  function handleCheck(todo: Todo) {
+    const checkTodo = { ...todo };
+
+    checkTodo.completed = !todo.completed;
+
+    function replaceTodo(prevState: State, patchedTodo: Todo) {
+      prevState.todos.splice(prevState.todos.indexOf(todo), 1, patchedTodo);
+
+      return prevState.todos;
+    }
+
+    patchTodo(checkTodo)
+      .then(patchedTodo => {
+        setState(prev => ({
+          ...prev,
+          todos: replaceTodo(prev, patchedTodo),
+        }));
+        removeLoading(checkTodo.id);
+      })
+      .catch(() => {
+        removeLoading(checkTodo.id);
+        setError(ErrorMessages.UpdateFail);
+      });
+  }
+
+  const handleCheckAll = () => {
+    const uncompletedTodo = state.todos.filter(todo => todo.completed === false);
+
+    if (uncompletedTodo.length > 0) {
+      setIsLoading(prev => (
+        [...prev, ...uncompletedTodo.map(todoId => todoId.id)]
+      ));
+
+      for (const todo of uncompletedTodo) {
+        handleCheck(todo);
+      }
+    } else {
+      setIsLoading(prev => (
+        [...prev, ...state.todos.map(todoId => todoId.id)]
+      ));
+
+      for (const todo of state.todos) {
+        handleCheck(todo);
+      }
+    }
+  };
+
+  function handleDeleteTodo(todo: Todo) {
+    deleteTodo(todo.id)
+      .then(() => {
+        setState((prev) => ({
+          ...prev,
+          todos: prev.todos.filter((t) => t.id !== todo.id),
+        }));
+        removeLoading(todo.id);
+        setIsEditing(null);
+      })
+      .catch(() => {
+        removeLoading(todo.id);
+        setError(ErrorMessages.DeleteFail);
+      });
+  }
+
+  function deleteAllCompleted() {
+    const completedTodo = state.todos.filter(todo => todo.completed);
+
+    setIsLoading(prev => (
+      [...prev, ...completedTodo.map(todoId => todoId.id)]
+    ));
+
+    for (const todo of completedTodo) {
+      handleDeleteTodo(todo);
+    }
+  }
+
+  function handleEditTodo(todo: Todo, newTitle: string) {
+    const editTodo = { ...todo };
+
+    editTodo.title = newTitle;
+
+    function replaceTodo(prevState: State, patchedTodo: Todo) {
+      prevState.todos.splice(prevState.todos.indexOf(todo), 1, patchedTodo);
+
+      return prevState.todos;
+    }
+
+    patchTodo(editTodo)
+      .then(patchedTodo => {
+        setState(prev => ({
+          ...prev,
+          todos: replaceTodo(prev, patchedTodo),
+        }));
+        removeLoading(editTodo.id);
+        setIsEditing(null);
+      })
+      .catch(() => {
+        removeLoading(editTodo.id);
+        setError(ErrorMessages.UpdateFail);
+      });
+  }
+
+  const contextValue = {
+    state,
+    setState,
+    handleCheck,
+    tempTodo,
+    handleDeleteTodo,
+    deleteAllCompleted,
+    setIsLoading,
+    isLoading,
+    handleEditTodo,
+    error,
+    isEditing,
+    setIsEditing,
+  };
```

Comment:
can you move and keep all this relevant logic (like contextValue etc) inside `TodoContext.ts`?

it will make your `app` file more readable